### PR TITLE
Implementation of dirty check for entity

### PR DESCRIPTION
**Pull request by dsnipe**
#170

### CODE REVIEW

File: lib/lotus/entity.rb [hunk 1]
```diff
@@ -144,13 +144,28 @@ def attributes(*attrs)
...
+      # If value new or changed, set it in hash changed_attributes.
+      #
+      # @params attr [Symbol] an attribute name
+      #
+      # @since 0.3.0
+      # @api private
+      def attr_accessor_fabric(attr)
+        attr_reader(attr)
+        define_method "#{attr}=" do |val|
+          _attribute_changed(attr) unless val == __send__(attr)
```
Comment (joneslee85) [inline lib/lotus/entity.rb]: What do you think about `if A != B`?

File: lib/lotus/entity.rb [hunk 2]
```diff
@@ -144,13 +144,28 @@ def attributes(*attrs)
           self.attributes.merge attrs
 
           attrs.each do |attr|
-            attr_accessor(attr) if define_attribute?(attr)
+            attr_accessor_fabric(attr) if define_attribute?(attr)
```
Comment (joneslee85) [inline lib/lotus/entity.rb]: What do you think about renaming it to `define_attr_accessor`?

File: lib/lotus/entity.rb [hunk 3]
```diff
@@ -144,13 +144,28 @@ def attributes(*attrs)
...
           @attributes ||= Set.new
         end
       end
 
+      # Create special setter method for attributes.
+      # If value new or changed, set it in hash changed_attributes.
+      #
+      # @params attr [Symbol] an attribute name
+      #
+      # @since 0.3.0
```
Comment (joneslee85) [inline lib/lotus/entity.rb]: Please change version to `x.x.x`

File: lib/lotus/entity.rb [hunk 4]
```diff
@@ -235,6 +252,46 @@ def update(attributes={})
...
+    #   @changed_attributes # => {:title => 'New title'}
+    #
+    def changed_attributes
+      @changed_attributes ||= Hash.new
+    end
+
+    # Return boolean of dirty state for model.
+    # If model changed, return true, else false.
+    #
+    # @since 0.3.0
```
Comment (joneslee85) [inline lib/lotus/entity.rb]: Please change version to `x.x.x`

File: lib/lotus/entity.rb [hunk 5]
```diff
@@ -235,6 +252,46 @@ def update(attributes={})
       end
     end
 
+    # Getter for hash of changed attributes.
+    # Return empty hash, if there is no changes
+    #
+    # @since 0.3.0
```
Comment (joneslee85) [inline lib/lotus/entity.rb]: Please change version to `x.x.x`

File: lib/lotus/entity.rb [hunk 6]
```diff
@@ -235,6 +252,46 @@ def update(attributes={})
...
+    # @api private
+    def _attribute_changed(attr)
+      @changed_attributes ||= Hash.new
+      @changed_attributes[attr] = __send__(attr)
+    end
+
+    # Clear all dirty data
+    #
+    # @since 0.3.0
```
Comment (joneslee85) [inline lib/lotus/entity.rb]: Please change version to `x.x.x`

File: lib/lotus/entity.rb [hunk 7]
```diff
@@ -235,6 +252,46 @@ def update(attributes={})
...
+      changed_attributes.size > 0
+    end
+
+    private
+
+    # Set changed attributes in Hash with their old values.
+    #
+    # @params attrs [Symbol] an attribute name
+    #
+    # @since 0.3.0
```
Comment (joneslee85) [inline lib/lotus/entity.rb]: Please change version to `x.x.x`

File: lib/lotus/entity.rb [hunk 8]
```diff
@@ -171,6 +186,7 @@ def inherited(subclass)
 
     # Defines a generic, inefficient initializer, in case that the attributes
     # weren't explicitly defined with `.attributes=`.
+    # Clear dirty data if id is set.
```
Comment (joneslee85) [inline lib/lotus/entity.rb]: Do you mean if the entity is persisted?

File: lib/lotus/entity.rb [hunk 9]
```diff
@@ -235,6 +252,46 @@ def update(attributes={})
       end
     end
 
+    # Getter for hash of changed attributes.
+    # Return empty hash, if there is no changes
```
Comment (joneslee85) [inline lib/lotus/entity.rb]: what do you think about following our conventional `@return [::Hash] the changed attributes`?

File: lib/lotus/entity.rb [hunk 10]
```diff
@@ -235,6 +252,46 @@ def update(attributes={})
...
+    #
+    # @example
+    #   @changed_attributes # => {:title => 'New title'}
+    #
+    def changed_attributes
+      @changed_attributes ||= Hash.new
+    end
+
+    # Return boolean of dirty state for model.
+    # If model changed, return true, else false.
```
Comment (joneslee85) [inline lib/lotus/entity.rb]: what do you think about following our conventional `@return [TrueClass, FalseClass] the changed changed indicator

File: test/entity_test.rb [hunk 11]
```diff
@@ -201,4 +201,20 @@ class Camera
...
   end
+
+  describe 'dirty checking' do
+
+    it 'dirty state when created' do
+      book = Book.new(title: 'Crime and Punishment', author: 'F. Dostoevskiy')
+      book.changed?.must_equal true
+    end
+
+    it 'change dirty state' do
```
Comment (joneslee85) [inline test/entity_test.rb]: `s\change\changes`

File: test/repository_test.rb [hunk 12]
```diff
@@ -115,6 +115,12 @@
           UserRepository.create(user1)
           user1.id.must_equal id
         end
+
+        it 'set model as not changed after persisted' do
```
Comment (joneslee85) [inline test/repository_test.rb]: `s\set\sets entity`

Comment (joneslee85) [inline test/repository_test.rb]: `s\not changed\unchanged`

File: test/repository_test.rb [hunk 13]
```diff
@@ -225,6 +231,13 @@ def to_int
           it "returns nil when the given id isn't associated with any entity" do
             UserRepository.find(1_000_000).must_be_nil
           end
+
+          it "not dirty when retrieved from persisted layer" do
```
Comment (joneslee85) [inline test/repository_test.rb]: `s\not\does not`

File: lib/lotus/entity.rb (hunk 2 above)
Comment (joneslee85) [inline lib/lotus/entity.rb]: on second look, `define_attribute?` is a poor naming, what do you think about renaming it to `defined_attribute?`

Comment (joneslee85) [inline lib/lotus/entity.rb]: Can you please also add additional `@examples` to cover the dirty tracking?

File: lib/lotus/entity.rb [hunk 14]
```diff
@@ -235,6 +252,46 @@ def update(attributes={})
       end
     end
 
+    # Getter for hash of changed attributes.
+    # Return empty hash, if there is no changes
+    #
+    # @since 0.3.0
+    #
+    # @example
+    #   @changed_attributes # => {:title => 'New title'}
```
Comment (joneslee85) [inline lib/lotus/entity.rb]: What do you think about amending the `@example` so that it include entity set/changes of attribute title? I believe it would help user to have a full picture of the workflow.

**Comment by joneslee85** [comment]
@dsnipe overall i am happy with the implementation
@jodosha love to hear your opinion on merging this feature or not

File: lib/lotus/entity.rb (hunk 2 above)
Comment (dsnipe) [inline lib/lotus/entity.rb]: no problem

File: lib/lotus/entity.rb (hunk 1 above)
Comment (dsnipe) [inline lib/lotus/entity.rb]: For simple conditions I used unless, but can change it to `!=`

File: lib/lotus/entity.rb (hunk 8 above)
Comment (dsnipe) [inline lib/lotus/entity.rb]: It only checks for the presence of ID, as keep track of where the model was initialized is not possible yet (from persisted level or by client)
So, if client set ID on init, it clear dirty data too.
Not best solution, but it's only way now to understand that model is persisted. Maybe you have ideas?

**Comment by dsnipe** [comment]
@joneslee85 done with fixes
will use #174 when merged

**Comment by dsnipe** [comment]
@joneslee85 I did rebase from master and used `persisted?` for checking
btw, did you seen my thoughts about `timestamps` realisation? What do you think?

**Comment by jodosha** [comment]
@dsnipe @joneslee85 Thanks for taking care of this. My opinion is that we're recreating AM/AR in Lotus::Model. While I find dirty tracking, timestamps, `#persisted?` useful features I'd not enable them by default.

I'd consider a different approach that consist in different modules like `Lotus::Entity::DirtyTracking`. These modules can be plugged in where needed.

Because this approach would be verbose, we can have a single module to include `Lotus::Entity::Persistence`, which will inject the extra features. We [already use this technique](https://github.com/lotus/helpers/blob/master/lib/lotus/helpers.rb#L20-L27) with `Lotus::Helpers`.

Likewise, this can be noisy too. My last suggestion is to introduce `Lotus::Model::Configuration#prepare_entities`, as a single place where to inject code for **all** the entities. We already use this for Controller and View.

**The bottom line is:** keep entities as simple as possible. This will guarantee speed for the default scenario. When more features are needed, devs can plug-in **one module** in **one place** and they are all set.

@dsnipe Can you please rework this as a module to inject after `Lotus::Entity`? Thank you very much for your effort.

File: lib/lotus/entity.rb [hunk 15]
```diff
@@ -235,6 +267,30 @@ def update(attributes={})
...
+    #     attributes :title
+    #   end
+    #
+    #   article = Article.new(title: 'The crime and punishment')
+    #   article.changed_attributes # => {}
+    #   article.title = 'Master and Margarita'
+    #   article.changed_attributes # => {:title => 'The crime and punishment'}
+    #
+    def changed_attributes
+      @changed_attributes ||= Hash.new
```
Comment (jodosha) [inline lib/lotus/entity.rb]: This isn't thread safe, this ivar should be set when the entity is initialized. Then the getter should be:

```ruby
def changed_attributes
  @changed_attributes.dup
end
```

In this way we prevent accidental modifications of the internal state of the object. Here an example:

```ruby
book.changed_attributes # => {:title => "The crime and punishment"}
book.changed_attributes.delete(:title)
book.changed_attributes # => {}
```

File: lib/lotus/entity.rb [hunk 16]
```diff
@@ -253,5 +309,36 @@ def update(attributes={})
...
+    private
+
+    # Set changed attributes in Hash with their old values.
+    #
+    # @params attrs [Symbol] an attribute name
+    #
+    # @since x.x.x
+    # @api private
+    def _attribute_changed(attr)
+      @changed_attributes ||= Hash.new
```
Comment (jodosha) [inline lib/lotus/entity.rb]: Again, this isn't thread safe.

**Comment by dsnipe** [comment]
@jodosha I had the same feeling about AM/AR, now I starting understand difference and your main idea. Thanks for the good explanation. 
I'll create new module `Lotus::Entity::DirtyTracking`, which can be included by developers in configuration block in future, if they need it. Now I using `persisted?` method. As I understand it will be in separate module, so better not depend on it?

**Comment by jodosha** [comment]
> Now I using persisted? method. As I understand it will be in separate module, so better not depend on it?

Yes. You can safely check for `.id` truthiness. (`if book.id` instead of `if book.persisted?`).

**Comment by dsnipe** [comment]
@jodosha, @joneslee85 
finished with module
can you review it? forgot to make rebase before push, will do it after review with fixes ;)

File: test/entity_test.rb [hunk 17]
```diff
@@ -201,4 +202,26 @@ class Camera
...
+    end
+
+    it 'changes dirty state' do
+      book1 = Book.new
+      book1.changed?.must_equal false
+      book1.title = 'War and Peace'
+      book1.changed?.must_equal true
+      book1.changed_attributes.must_include :title
+    end
+
```
Comment (jodosha) [inline test/entity_test.rb]: Can you please add another test for `#update`? We expect it to already work because it uses internally `public_send("#{attribute}=", value)`. But if we change that implementation, dirty tracking won't work anymore. We need a test to prevent regressions.

File: lib/lotus/entity/dirty_tracking.rb [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+        # @params attr [Symbol] an attribute name
+        #
+        # @since x.x.x
+        #
+        # @see Lotus::Entity
+        #
+        # @api private
+        def define_attr_accessor(attr)
+          attr_reader(attr)
+          define_method "#{attr}=" do |val|
```
Comment (jodosha) [inline lib/lotus/entity/dirty_tracking.rb]: Can you please benchmark this implementation and compare it with the following?

```ruby
class_eval %{
  def #{ attr }=(value)
    _attribute_changed(:#{ attr }) if value != @#{ attr }
    @#{ attr } = value
  end
}
```

File: lib/lotus/entity.rb [hunk 19]
```diff
@@ -138,24 +138,35 @@ module ClassMethods
...
       #   DeletedUser.attributes => #<Set: {:id, :name, :deleted_at}>
+      #
       def attributes(*attrs)
         if attrs.any?
           attrs = Lotus::Utils::Kernel.Array(attrs)
           self.attributes.merge attrs
 
           attrs.each do |attr|
-            attr_accessor(attr) if define_attribute?(attr)
+            define_attr_accessor(attr) if defined_attribute?(attr)
```
Comment (mvader) [inline lib/lotus/entity.rb]: Why wrap ```attr_accessor(attr)``` inside ```define_attr_accessor(attr)```? ```define_attr_accessor``` ends up being exactly like ```attr_accessor```. Why not use just ```attr_accessor```?

Comment (dsnipe) [inline lib/lotus/entity.rb]: @mvader because it's easier to override than behaviour; when you inject another module.
you can see it there: https://github.com/dsnipe/model/blob/ac9bcd5ac2612371c214177ae7ab4c5c9598b5ee/lib/lotus/entity/dirty_tracking.rb#L45-L53

Comment (dsnipe) [inline lib/lotus/entity.rb]: I think I need add better explanation in the comments

**Comment by joneslee85** [comment]
@dsnipe can you please rebase?

**Comment by dsnipe** [comment]
@joneslee85 done

**Comment by AlfonsoUceda** [comment]
@dsnipe you have a lot of tests failing, can you take a look?

**Comment by dsnipe** [comment]
@AlfonsoUceda yep, was bad rebase. 
Now all green :)

File: lib/lotus/entity/dirty_tracking.rb [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+      # Override initialize process.
+      # Init @changed_attributes and Clear dirty data if
+      # entity persisted.
+      #
+      # @param attributes [Hash] a set of attribute names and values
+      #
+      # @since x.x.x
+      #
+      # @see Lotus::Entity#initialize
+      #
```
Comment (AlfonsoUceda) [inline lib/lotus/entity/dirty_tracking.rb]: remove this line

File: lib/lotus/entity/dirty_tracking.rb [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+      #     include Lotus::Entity::DirtyTracking
+      #
+      #     attributes :title
+      #   end
+      #
+      #   article = Article.new(title: 'The crime and punishment')
+      #   article.changed_attributes # => {}
+      #   article.title = 'Master and Margarita'
+      #   article.changed_attributes # => {:title => 'The crime and punishment'}
+      #
```
Comment (AlfonsoUceda) [inline lib/lotus/entity/dirty_tracking.rb]: remove this line

File: lib/lotus/entity/dirty_tracking.rb [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+      def changed_attributes
+        @changed_attributes.dup
+      end
+
+      # Return boolean of dirty state for model.
+      #
+      # @return [TrueClass, FalseClass] the changed indicator
+      #
+      # @since x.x.x
+      #
```
Comment (AlfonsoUceda) [inline lib/lotus/entity/dirty_tracking.rb]: remove this line

File: lib/lotus/entity/dirty_tracking.rb [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+      end
+
+      # Clear all information about dirty data
+      #
+      # @since x.x.x
+      #
+      def _clear_changes_information
+        @changed_attributes = {}
+      end
+
```
Comment (AlfonsoUceda) [inline lib/lotus/entity/dirty_tracking.rb]: remove this line

**Comment by AlfonsoUceda** [comment]
:+1:

**Comment by joneslee85** [comment]
:+1:

**Comment by jodosha** [comment]
@dsnipe Thank you, just merged :+1:
